### PR TITLE
[RFC][libc++] Reworks clang-tidy selection.

### PR DESCRIPTION
**Pull request by mordante**
The current selection is done in the test scripts which gives the user no control where to find clang-tidy. The version selection itself is hard-coded and not based on the version of clang used.

This moves the selection to configuration time and tries to find a better match.
- Mixing the version of clang-tidy and the clang libraries causes ODR violations. This results in practice in crashes or incorrect results.
- Mixing the version of clang-tidy and the clang binary sometimes causes issues with supported diagnostic flags. For example, flags tested against clang 17 may not be available in clang-tidy 16.

Currently clang-tidy 18.1 can be used, it tests against the clang libraries version 18. This is caused by the new LLVM version numbering scheme.

The new selection tries to match the clang version or the version of the HEAD and the last 3 releases. (During the release period libc++ supports 4 versions instead of the typical 3 versions.)

### CODE REVIEW

File: libcxx/utils/libcxx/test/features.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 # ===----------------------------------------------------------------------===##
 
 from libcxx.test.dsl import *
+import libcxx.test.config as config
```
Comment (ldionne) [inline libcxx/utils/libcxx/test/features.py:10]: I really like how cleaner the logic is after this patch, however I would really like to avoid threading this through the CMake. Ideally, the test suite would be the one to figure out the executable to use for clang-tidy. Something like this:

```diff
diff --git a/libcxx/utils/libcxx/test/params.py b/libcxx/utils/libcxx/test/params.py
index 89d9f22e9dc6..c6b32f6fae61 100644
--- a/libcxx/utils/libcxx/test/params.py
+++ b/libcxx/utils/libcxx/test/params.py
@@ -366,6 +366,16 @@ DEFAULT_PARAMETERS = [
         default=f"{shlex.quote(sys.executable)} {shlex.quote(str(Path(__file__).resolve().parent.parent.parent / 'run.py'))}",
         help="Custom executor to use instead of the configured default.",
         actions=lambda executor: [AddSubstitution("%{executor}", executor)],
+    ),
+    Parameter(
+        name='clang-tidy-executable',
+        type=str,
+        default=_getSuitableClangTidy(),
+        help="TODO",
+        actions=lambda exe: [] if exe is None else [
+            AddSubstitution('%{clang-tidy}', exe),
+            AddFeature('has-clang-tidy')
+        ]
     )
 ]
 # fmt: on
```

File: cmake/Modules/LLVMVersion.cmake [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# The LLVM Version number information
+
+if(NOT DEFINED LLVM_VERSION_MAJOR)
```
Comment (ldionne) [inline cmake/Modules/LLVMVersion.cmake:3]: I think this is worthwhile as its own change, and in particular we could probably derive `_LIBCPP_VERSION` from it, which would simplify the job of the release manager.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/utils/libcxx/test/features.py (hunk 1 above)
Comment (mordante) [reply libcxx/utils/libcxx/test/features.py:10]: I've given this a bit more thought. One of the reasons to move it to CMake was to be able to select the proper clang library to build the plugin against. Moving this to features re-introduces that issue. Let's discuss this next time we meet.

File: libcxx/test/tools/clang_tidy_checks/CMakeLists.txt [hunk 3]
```diff
@@ -5,10 +5,10 @@
 set(LLVM_DIR_SAVE ${LLVM_DIR})
 set(Clang_DIR_SAVE ${Clang_DIR})
 
-find_package(Clang 18)
-if (NOT Clang_FOUND)
-  find_package(Clang 17)
-endif()
+# Since the Clang C++ ABI is not stable the Clang libraries and clang-tidy
+# version must match. Otherwise there likely will be ODR-violations. This had
```
Comment (ldionne) [inline libcxx/test/tools/clang_tidy_checks/CMakeLists.txt:9]:
```suggestion
# versions must match. Otherwise there likely will be ODR-violations. This had
```

File: libcxx/utils/libcxx/test/params.py [hunk 4]
```diff
@@ -53,17 +53,14 @@
     "-Wunused-parameter",
     "-Wunreachable-code",
     "-Wno-unused-local-typedef",
-
```
Comment (ldionne) [inline libcxx/utils/libcxx/test/params.py:56]: These removals seem unrelated.

File: libcxx/utils/libcxx/test/params.py [hunk 5]
```diff
@@ -366,6 +394,16 @@ def getSizeOptimizationFlag(cfg):
         default=f"{shlex.quote(sys.executable)} {shlex.quote(str(Path(__file__).resolve().parent.parent.parent / 'run.py'))}",
         help="Custom executor to use instead of the configured default.",
         actions=lambda executor: [AddSubstitution("%{executor}", executor)],
-    )
+    ),
+    Parameter(
+        name='clang-tidy-executable',
+        type=str,
+        default=lambda cfg: getSuitableClangTidy(cfg),
+        help="TODO",
```
Comment (ldionne) [inline libcxx/utils/libcxx/test/params.py:402]: Leftover!

**Review by ldionne — APPROVED** [review]
I like this! I have minor comments but let's ship this. We should separate the LLVM version setting out of this patch and it would be nice to remove `LIBCXX_ENABLE_CLANG_TIDY` as a follow-up too.

File: cmake/Modules/LLVMVersion.cmake (hunk 2 above)
Comment (mordante) [reply cmake/Modules/LLVMVersion.cmake]: This is now https://github.com/llvm/llvm-project/pull/84641
